### ISSUE TITLE
Mutex::isAcquired() is not working if is called in different process

### ISSUE THREAD

**Issue by goncaloe** [yiisoft/yii2#18611]
### What steps will reproduce the problem?
Define 2 actions in console Controller:
`
    public function actionA(){
        $mutexName = 'test_mutex';
        $mutex = \Yii::createObject([
            'class' => \yii\mutex\FileMutex::class,
            'mutexPath' => '@runtime/mutex',
        ]);

        echo "acquire mutex\n";
        $mutex->acquire($mutexName);

        sleep(9999);

        echo "release mutex\n";
        $mutex->release($mutexName);
    }

    public function actionB(){
        $mutexName = 'test_mutex';
        $mutex = \Yii::createObject([
            'class' => \yii\mutex\FileMutex::class,
            'mutexPath' => '@runtime/mutex',
        ]);

        if($mutex->isAcquired($mutexName)){
            echo "mutex is acquired\n";
        }
        else {
            echo "mutex is not acquired\n";
        }
    }
`

Open a console window and execute (let it in running):
$ yii test/a
acquire mutex

Open a second console window and execute:
$ yii test/b
mutex is not acquired

### What is the expected result?
On second window is expected the output: mutex is acquired

### What do you get instead?
On second window: mutex is not acquired
Tested with DbMutex with same result

### Additional info

| Q                | A
| ---------------- | ---
| Yii version      | 2.0.41
| PHP version      | 7.3.9
| Operating system | Windows 10

**Comment by bizley** [comment]
Is `$mutex->acquire($mutexName);` in A returning `true`?

**Comment by goncaloe** [comment]
Yes, returns true;
The problem is in Mutex::isAcquired() is not checking the file / database of mutex, but only check the `$this->_locks` (that is not filled in second request)

**Comment by ivan-redooc** [comment]
@goncaloe it seem to me because in actionB you don't follow the pattern "try & check", but you go directly to the check, eg:

```PHP
public function actionB(){
    $mutexName = 'test_mutex';
    $mutex = \Yii::createObject([
        'class' => \yii\mutex\FileMutex::class,
        'mutexPath' => '@runtime/mutex',
    ]);
    
    // try
    $mutex->acquire($mutexName);
    
    // check
    if($mutex->isAcquired($mutexName)){
        echo "mutex is acquired\n";
    }
    else {
        echo "mutex is not acquired\n";
    }
}

```

**Comment by bizley** [comment]
Yes, but idea of mutex is to be used in different processes as well.

**Comment by ivan-redooc** [comment]
> Yes, but idea of mutex is to be used in different processes as well.

Of course, but the [documentation](https://www.yiiframework.com/doc/api/2.0/yii-mutex-mutex) show the pattern to use:

> The Mutex component allows mutual execution of concurrent processes in order to prevent "race conditions".
> 
> This is achieved by using a "lock" mechanism. Each possibly concurrent thread cooperates by acquiring a lock before accessing the corresponding data.

```PHP
if ($mutex->acquire($mutexName)) {
    // business logic execution
} else {
    // execution is blocked!
}
```

And for I understand, the correct approach is **to try** to _acquire_ the mutex **to check** if the mutex is available.

IMHO the real purpose of _isAcquired_ is the flow control in the **same** process **after** tryed to _acquire_ the mutex, with or without success.

**Comment by goncaloe** [comment]
@ivan-redooc In my case, I have to check in web app if a console action (synchronization that takes a long time) is running (mutex is locked).
If it is unlocked, I show a button to start synchronization.

This should be possible, without using the acquire() function.

**Comment by ivan-redooc** [comment]
> This should be possible, without using the acquire() function.
I don't agree, it's the lock pattern, you cannot check without lock.  

So here the obvious solution:
```PHP
if ($mutex->acquire($mutexName)) {
    $mutex->release($mutexName);
  // show the button
} else {
    // show the message "still working"
}
```

In other environments, with strong concurrency, you can simply always show the button because you will don't know if the mutex will be free when you click it.

**Comment by mikehaertl** [comment]
@ivan-redooc Why would you ever want to check for `isAcquired()` in the same process where you acquired the lock? As you said the recommended pattern is:

```
if ($mutex->acquire($key)) {
    // Got mutex here. It makes no sense to check for isAcquired() again.
} else {
    // no mutex
}
```

That's also the pattern I use. I've never used `isAcquired()`. But I would have expected that I could use it wherever I want not just in the same process that acquired the lock. So I would say that the issue is valid.

**Comment by mikehaertl** [comment]
I see that `isAcquired()` was just recently added in #18151. To be honest I'm not sure if this was such a good idea. As we see in this issue it gives the wrong impression that you can use it everywhere. While in fact it's only useful in the same process where you acquired the lock. But mutexes whole point is to work across multiple processes. 

The real fix would be to add a `abstract protected checkLock()` method that child classes must implement. But IMO that's not an option either as it could break BC (and maybe not every backend supports such a check without obtaining the lock).

The way I see it: There is no fix. So we should put a big warning message above `isAcquired()` to make it clear that it does *not* work in every scenario. Maybe even mark it as deprecated.

**Comment by bizley** [comment]
Hm, why not fix `isAcquired` to not check its internal `_locks` state but rather to directly check the storage it uses? Why cannot it be fixed?

**Comment by mikehaertl** [comment]
@bizley How would you do that? We only have `abstract acquireLock()` and `abstract releaseLock()`. But we don't want to acquire the lock, we only want to check it. So we would need `abstract checkLock()`. This is

1. BC break (what if someone implemented their own Mutex child class?)
2. not every backend supports it. E.g. with Mysql you can't check if a lock is obtained. At least not easily.

**Comment by bizley** [comment]
No, there is `Mutex::isAcquired()` method that is fine for the abstraction but should and must be extended for implementations. There we can do the actual check.

**Comment by mikehaertl** [comment]
But point 2 still holds true: You can't check for a lock in MySQL without obtaining the lock (which is not what you want wenn you call `isAcquired()` - otherwhise you could also call `acquire()`).

And if you implemented your own child class, you may have not implementation for `isAcquired()` - and so the situation in this issue can still hit you.

**Comment by bizley** [comment]
I can see there is something like IS_USED_LOCK on MySQL but I've never used it before.

You are right about this custom implementations but if we can write the warning on this method at least someone can read it and realize it. And we can fix the core implementations.

**Comment by mikehaertl** [comment]
Ups, right, I missed that. `IS_USED_LOCK()` (or `IS_FREE_LOCK()`) should work indeed. But this must also be checked for the other DBMS we have mutexes for.

**Comment by My6UoT9** [comment]
I expect that isAcquired does what it does. If i did not acquire a mutex in the current process, then isAcquired should return false.
Maybe I interpret the english incorrect, but that is what I understand from the wording, Mutex Is Acquired.

So maybe a secondary function, Mutex isLocked()

**Comment by mikehaertl** [comment]
I'm afraid that if we change this, it will break the initial use case that @rhertogh reported in #18151 

> It would be used in a model's save function to ensure it is only called when the mutex is active. The mutex itself would be acquired in different places and wraps different models.

Consider this code:

```php
public function beforeSave()
{
    if (!Yii::$app->mutex->acquire('somekey')) {
        return false;
    }
}

public function afterSave()
{
    if (Yii::$app->mutex->isAcquired('somekey')) {
        // do dangerous things here....
    }
}
```

 * Current situation: Code in `afterSave()` will only be executed, if the lock was acquired **in the same process** (which happened most likely in `beforeSave()` - but that's not even guaranteed to be the case)
 * After the proposed change: Code in `afterSave()` will even be executed if the lock was acquired **by another process in another instance of the active record**. 

I actually think, that @rhertogh 's problem should rather be solved in his own code, like:


```php
public $hasLock = false;

public function beforeSave()
{
    if (!Yii::$app->mutex->acquire('somekey')) {
        return false;
    } else {
        $this->hasLock = true;
    }
}

public function afterSave()
{
    if ($this->hasLock) {
        // do dangerous things here....
    }
}
```

**Comment by ivan-redooc** [comment]
@mikehaertl I prefer 

```PHP
    if (Yii::$app->mutex->isAcquired('somekey')) {
        // do dangerous things here....
    }
```

to 

```PHP
    if ($this->hasLock) {
        // do dangerous things here....
    }

```
because the Mutex keeps the responsability to know own status. 

Of course an improve to the documentation of _isAcquired_ should be necessary.

**Comment by mikehaertl** [comment]
@ivan-redooc That's not the point. It seems there's still a lot confusion here. Let me try again to explain the two use cases as I see it:

 1. Case in #18151: `isAcquired()` answers the question "Do **we** have the lock?" I.e. does the current instance  of ActiveRecord in the current PHP process have obtained the lock? That's what the implementation right now does. A parallel process gets `false` - even though the lock is not available!
 2. After the fix requested in this issue: `isAcquired()` answers the question "Does **any** PHP process have the lock?". That's what mutex is made for.

That's something completely different. If we fix the issue here, the behavior will change from 1) to 2).

The code I showed above is a workaround for those that rely on behavior 1). It's not a choice. The current implementation is simply flawed and does not get the point of a mutex.

**Comment by samdark** [comment]
I think the right solution here is to emphasize current method behavior.

**Comment by samdark** [comment]
@bizley because fixing it will definitely cause backwards compatibility problems for original use-case.

**Comment by bizley** [comment]
Maybe, but the original one is very limited, so now we don't have proper mutex in place. Maybe I will be able to fix it.

**Comment by samdark** [comment]
@bizley we do have proper mutex and it should be used without calling `isAcquired()`.

**Comment by bizley** [comment]
Yes, it was a bit too harsh, sorry. Ok, then, all good.

**Comment by rob006** [comment]
Current implementation of `isAcquired()` is the only implementation that make sense. You can't check whether other processes acquired lock, because result may be outdated right after method call (`$mutex->isAcquired($mutexName) && $mutex->isAcquired($mutexName)` could return `false` if lock was acquired/release by other process between these calls). Trying to include locks acquired by other processes would make this method literally useless, since it result would be unreliable by design. 

```php
if (!$mutex->isAcquired($mutexName)) {
    // I don't know if lock is acquired, because previous result of `$mutex->isAcquired($mutexName)` is already outdated
}

**Comment by samdark** [comment]
Well, yes. It won't be atomic in this case 👍

**Comment by rhertogh** [comment]
@goncaloe As @rob006 has described a mutex is to prevent concurrent execution that would cause problems (more info: https://en.wikipedia.org/wiki/Mutual_exclusion)
It seems that you are trying to get the "status" of another task that is running. If that is the case it would be better to store the task status, e.g. in a database. Perhaps yiisoft/yii2-queue suits your needs.
